File: module-app/src/main/java/maple/expectation/service/v2/policy/TableBasedCostStrategy.java
package maple.expectation.service.v2.policy;

import java.util.EnumMap;
import java.util.Map;
import java.util.TreeMap;
import maple.expectation.domain.v2.CubeType;
import maple.expectation.domain.v2.PotentialGrade;
import org.springframework.stereotype.Component;

/**
 * 테이블 기반 비용 계산 전략 (기본 구현)
 *
 * <p>책임 (Single Responsibility Principle):
 *
 * <ul>
 *   <li>정적 테이블에서 비용 조회
 *   <li>EnumMap을 활용한 O(1) 조회 성능
 * </ul>
 *
 * <p>OCP 준수:
 *
 * <ul>
 *   <li>확장: {@link CostCalculationStrategy} 인터페이스 구현
 *   <li>수정 방지: 새로운 큐브 타입 추가 시 기존 코드 수정 불필요
 * </ul>
 */
@Component
public class TableBasedCostStrategy implements CostCalculationStrategy {

  private final Map<
          CubeType, TreeMap<Integer, EnumMap<maple.expectation.domain.v2.PotentialGrade, Long>>>
      costMasterTable;

  public TableBasedCostStrategy() {
    this.costMasterTable = initializeCostTable();
  }

  @Override
  public long calculateCost(CubeType type, int level, String grade) {
    // Fail-Fast: 잘못된 입력 즉시 예외 (Silent Failure 방지)
    maple.expectation.domain.v2.PotentialGrade validGrade =
        maple.expectation.domain.v2.PotentialGrade.valueOf(grade.toUpperCase());

    TreeMap<Integer, EnumMap<PotentialGrade, Long>> typeTable = costMasterTable.get(type);
    if (typeTable == null) {
      throw new IllegalStateException("Unknown CubeType: " + type);
    }

    // 레벨에 맞는 가장 가까운 하위 구간 키를 찾음 (예: 210레벨 -> 200 키 선택)
    Integer levelKey = typeTable.floorKey(level);
    if (levelKey == null) {
      levelKey = typeTable.firstKey();
    }

    // O(1) EnumMap lookup
    return typeTable.get(levelKey).get(validGrade);
  }

  /**
   * 비용 테이블 초기화 (생성자에서 호출)
   *
   * @return 비용 마스터 테이블
   */
  private Map<CubeType, TreeMap<Integer, EnumMap<maple.expectation.domain.v2.PotentialGrade, Long>>>
      initializeCostTable() {
    Map<CubeType, TreeMap<Integer, EnumMap<maple.expectation.domain.v2.PotentialGrade, Long>>>
        table = new EnumMap<>(CubeType.class);

    // 1. 블랙큐브(윗잠재) 비용 테이블
    TreeMap<Integer, EnumMap<PotentialGrade, Long>> blackTable = new TreeMap<>();
    blackTable.put(0, createGradeMap(4000000L, 16000000L, 34000000L, 40000000L));
    blackTable.put(160, createGradeMap(4250000L, 17000000L, 36125000L, 42500000L));
    blackTable.put(200, createGradeMap(4500000L, 18000000L, 38250000L, 45000000L));
    blackTable.put(250, createGradeMap(5000000L, 20000000L, 42500000L, 50000000L));
    table.put(CubeType.BLACK, blackTable);

    // 2. 에디셔널큐브(아랫잠재) 비용 테이블
    TreeMap<Integer, EnumMap<PotentialGrade, Long>> addiTable = new TreeMap<>();
    addiTable.put(0, createGradeMap(13000000L, 36400000L, 44200000L, 52000000L));
    addiTable.put(160, createGradeMap(13812500L, 38675000L, 46962500L, 55250000L));
    addiTable.put(200, createGradeMap(14625000L, 40950000L, 49725000L, 58500000L));
    addiTable.put(250, createGradeMap(16250000L, 45500000L, 55250000L, 65000000L));
    table.put(CubeType.ADDITIONAL, addiTable);

    // 3. 레드큐브 - 레벨 구분 없음, 비용 계산 시 1을 반환하여 '개수'로 취급 가능
    TreeMap<Integer, EnumMap<PotentialGrade, Long>> redTable = new TreeMap<>();
    redTable.put(0, createGradeMap(1L, 1L, 1L, 1L));
    table.put(CubeType.RED, redTable);

    return table;
  }

  /** EnumMap 생성 헬퍼 - RARE, EPIC, UNIQUE, LEGENDARY 순서로 비용 매핑 */
  private EnumMap<maple.expectation.domain.v2.PotentialGrade, Long> createGradeMap(
      long rare, long epic, long unique, long legendary) {
    EnumMap<maple.expectation.domain.v2.PotentialGrade, Long> map =
        new EnumMap<>(maple.expectation.domain.v2.PotentialGrade.class);
    map.put(maple.expectation.domain.v2.PotentialGrade.RARE, rare);
    map.put(maple.expectation.domain.v2.PotentialGrade.EPIC, epic);
    map.put(maple.expectation.domain.v2.PotentialGrade.UNIQUE, unique);
    map.put(maple.expectation.domain.v2.PotentialGrade.LEGENDARY, legendary);
    return map;
  }
}

File: module-core/src/main/kotlin/maple/expectation/core/policy/TableBasedCostStrategy.kt
package maple.expectation.core.policy

import java.util.EnumMap
import java.util.NavigableMap
import java.util.TreeMap
import maple.expectation.core.domain.model.CubeType
import maple.expectation.core.domain.model.PotentialGrade

/**
 * Table-based cost calculation strategy (default implementation).
 *
 * <p>Responsibilities (Single Responsibility Principle):
 * - Lookup cost from static table
 * - O(1) lookup performance using EnumMap
 *
 * <p>OCP Compliance:
 * - Extension: Implements {@link CostCalculationStrategy} interface
 * - Modification prevention: New cube types added without code changes
 */
class TableBasedCostStrategy : CostCalculationStrategy {

    private val costMasterTable: Map<CubeType, NavigableMap<Int, EnumMap<PotentialGrade, Long>>>

    init {
        this.costMasterTable = initializeCostTable()
    }

    override fun calculateCost(type: CubeType, level: Int, grade: String): Long {
        // Fail-Fast: Immediate exception on invalid input (prevent Silent Failure)
        val validGrade = PotentialGrade.fromKorean(grade)

        val typeTable = costMasterTable[type]
            ?: throw IllegalStateException("Unknown CubeType: $type")

        // Find closest lower bracket key for level (e.g., level 210 -> key 200)
        val levelKey = typeTable.floorKey(level) ?: typeTable.firstKey()

        // O(1) EnumMap lookup
        return typeTable[levelKey]!![validGrade]!!
    }

    /**
     * Initialize cost table (called from constructor).
     *
     * @return Cost master table
     */
    private fun initializeCostTable(): Map<CubeType, NavigableMap<Int, EnumMap<PotentialGrade, Long>>> {
        val table = mutableMapOf<CubeType, NavigableMap<Int, EnumMap<PotentialGrade, Long>>>()

        // 1. Black Cube (upper potential) cost table
        val blackTable: NavigableMap<Int, EnumMap<PotentialGrade, Long>> = TreeMap<Int, EnumMap<PotentialGrade, Long>>().apply {
            put(0, createGradeMap(4_000_000L, 16_000_000L, 34_000_000L, 40_000_000L))
            put(160, createGradeMap(4_250_000L, 17_000_000L, 36_125_000L, 42_500_000L))
            put(200, createGradeMap(4_500_000L, 18_000_000L, 38_250_000L, 45_000_000L))
            put(250, createGradeMap(5_000_000L, 20_000_000L, 42_500_000L, 50_000_000L))
        }
        table[CubeType.BLACK] = blackTable

        // 2. Additional Cube (lower potential) cost table
        val addiTable: NavigableMap<Int, EnumMap<PotentialGrade, Long>> = TreeMap<Int, EnumMap<PotentialGrade, Long>>().apply {
            put(0, createGradeMap(13_000_000L, 36_400_000L, 44_200_000L, 52_000_000L))
            put(160, createGradeMap(13_812_500L, 38_675_000L, 46_962_500L, 55_250_000L))
            put(200, createGradeMap(14_625_000L, 40_950_000L, 49_725_000L, 58_500_000L))
            put(250, createGradeMap(16_250_000L, 45_500_000L, 55_250_000L, 65_000_000L))
        }
        table[CubeType.ADDITIONAL] = addiTable

        // 3. Red Cube - no level distinction, returns 1 to treat as 'count'
        val redTable: NavigableMap<Int, EnumMap<PotentialGrade, Long>> = TreeMap<Int, EnumMap<PotentialGrade, Long>>().apply {
            put(0, createGradeMap(1L, 1L, 1L, 1L))
        }
        table[CubeType.RED] = redTable

        return table
    }

    /**
     * Create EnumMap helper - maps costs in RARE, EPIC, UNIQUE, LEGENDARY order.
     */
    private fun createGradeMap(
        rare: Long,
        epic: Long,
        unique: Long,
        legendary: Long,
    ): EnumMap<PotentialGrade, Long> {
        val map = EnumMap<PotentialGrade, Long>(PotentialGrade::class.java)
        map[PotentialGrade.RARE] = rare
        map[PotentialGrade.EPIC] = epic
        map[PotentialGrade.UNIQUE] = unique
        map[PotentialGrade.LEGENDARY] = legendary
        return map
    }
}
